### PR TITLE
Introduce new, vanilla-compatible server info protocol (also known as: the solution to all problems)

### LINKED ISSUE #130: Server info not necessarily sorted at the right time
The code assumes that the packet containing the last players arrives last.

CC @fstd

### PR DESCRIPTION
**Pull request by heinrich5991**
This means that we have a reliable and fast way to query for extended info,
while also not wasting network bandwidth.

The protocol is designed to be extensible, there's four bytes space for
encoding more request types (currently zeroed), and there's one string in each
response packet and one string for each player available (currently the empty
string).

The protocol itself has no problems with more than 64 players, although the
current client implementation will drop the player info after the 64th player,
because it uses a static array for storage.

Also fixes #130, the player list is just sorted each time new player info
arrives.

### CODE REVIEW

**Comment by heinrich5991** [comment]
Please review, but not merge yet, so we can create a forward-compatible protocol this time! I'm awaiting your comments.

**Comment by heinrich5991** [comment]
@fstd suggests to use `xe` instead of `ex` as identifying bytes, because it is less obvious. /agree

**Comment by heinrich5991** [comment]
NOTE: This also doesn't send client names to clients in the server info sent during the game.

**Comment by heinrich5991** [comment]
Perhaps we might want to include the map CRC in the info, so we could show whether the client needs to download a map?

**Comment by east** [comment]
Yes, would be nice. Also once we talked about providing a custom http download url in server info to support hosters not part of ddnet. Wouldn't this be the right time for implementation?

**Comment by heinrich5991** [comment]
In the following document, "vanilla" refers to unmodified Teeworlds.


Extended server info request
============================

An extended server info can be requested from the Teeworlds server by sending
the following UDP packet:

    [2] magic_bytes
    [4] reserved
    [4] padding
    [4] vanilla_request
    [1] token

`magic_bytes` must be the ASCII representation of "ex".

`reserved` is reserved for future protocol versions and must be zeroed by the
sender and ignored by the receiver with this protocol version.

`padding` must be filled with 0xff bytes.

`vanilla_request` must be the ASCII representation of "gie3".

`token` is an arbitrary byte chosen by the client which will be sent back with
the server info.

If a server receives such a request (detected by the `magic_bytes` field), it
can either respond with the vanilla server info, or by sending the extended
server info, described in the following.

If the server does not recognize the `magic_bytes`, it should send back the
vanilla server info.


Extended server info
====================

The extended server info can consist of several packets, the "main" packet and
the "more" packets, which can contain additional player data in case they don't
fit in the first packet.

The `str` type is a zero-terminated UTF-8 string. Implementations should be
prepared to either replace invalid UTF-8 with the replacement character or to
treat these fields as opaque zero-terminated byte sequences.

The `int` type is an integer encoded in the decimal system in a zero-terminated
ASCII string.

NOTE: The reference implementation clears control characters (replacing bytes
with a value of less than 32 with the ASCII space ' ') and skips leading
whitespace in all strings.

The receiver of such a server info can detect whether it has received the
server info completely by comparing the number of players received with the
number of players set in the main packet.


The main packet
---------------

The main server info packet looks like follows:

    [ 10] padding
    [  4] type
    [int] token
    [str] name
    [str] map
    [str] game_type
    [int] flags
    [int] num_players
    [int] max_players
    [int] num_clients
    [int] max_clients
    [str] reserved
    [  *] players

`padding` must be filled with 0xff bytes.

`type` must be the ASCII representation of "iext".

`token` must be the token from the request. If the server info isn't sent upon
a request, it mut be -1.

`name`, `map`, `game_type` are the server's name, its current map and gametype,
respectively.

There is currently only one flag for the `flags` field, namely the password
flag, whose value is 1. It must be set if entering the server requires a
password, and unset in the other case.

`num_players`, `max_players`, `num_clients`, `max_clients`: These values
describe how many players (people that are not in "spectator mode") and clients
(people that are connected to the server) the server currently has (`num_*`)
and is able to handle (`max_*`). The following natural inequalities must hold:

                     <= max_players
    0 <= num_players                <= max_clients
                     <= num_clients

Note that there is no upper limit for `max_clients` enforced by the protocol.

The field `reserved` is reserved for future protocol versions. Senders of this
message must set the field to the empty string, and receivers of this message
must ignore its value.

`players` contains the player info described further below.


The "more" packet
-----------------

A "more" packet looks like this:

    [ 10] padding
    [  4] type
    [int] token
    [int] packet_no
    [str] reserved
    [  *] players

`padding`, `token`, `reserved`, `players`: See the explanation on the main
packet.

`type` must contain the ASCII representation of "iex+".

`packet_no` must be packet number in the server info response. It must be
greater than 0 and less than 64. The main packet has an implicit `packet_no` of
0. This field can be used to detect duplicated packets.


Player info
-----------

A single player info is defined as:

    [str] name
    [str] clan
    [int] country
    [int] score
    [int] is_player
    [str] reserved

The `name`, `clan` and `score` field contain the player's name, clan and score,
respectively.

The `country` field contains the ISO 3166-1 numeric code of the player's
country, or -1 if unset.

The `is_player` field contains a non-zero integer if the player is not a
spectator, and 0 otherwise. Producers of this message should always use 1 to
indicate a non-spectator client.

**Comment by heinrich5991** [comment]
@fstd Here's your documentation.

**Comment by heinrich5991** [comment]
@east I believe that's better handled as part of the `NETMSG_MAP_CHANGE` message.

File: src/engine/client/client.cpp [hunk 1]
```diff
@@ -1346,135 +1346,197 @@ void CClient::ProcessConnlessPacket(CNetChunk *pPacket)
...
 				m_CurrentServerInfoRequestTime = -1;
 			}
 		}
+
+		if(SavedType == SERVERINFO_VANILLA && Is64Player(&Info) && pEntry)
+		{
+			pEntry->m_Request64Legacy = true;
+			// Force a quick update.
+			m_ServerBrowser.RequestImpl64(pEntry->m_Addr, pEntry);
+		}
```
Comment (heinrich5991) [inline src/engine/client/client.cpp]: Mh. This should probably only trigger if we don't already have a 64p server info.

**Comment by heinrich5991** [comment]
Added map size and CRC to the response packet.

@east How would such a challenge response system look like?

**Comment by east** [comment]
This is what I thought about:

When the server receives an extended request, it can optionally initiate a token challenge by sending a message containing an ip:port specific token + the one byte token from the client. The client receives the message and verifies its own token to make sure it was sent from the right source (1 byte is a bit unsafe though :/). Then it sends another request containing the generated token of the server.

In short:
server -> client
 - new message type e.g. 'RETRY_WITH_TOKEN' which used to send a generated token + client token

client -> server
 - request needs to be extended by the generated token

**Comment by timgame** [comment]
Couldn't this be a solution for #344?

**Comment by heinrich5991** [comment]
That wasn't the problem with #344.

**Comment by PathosEthosLogos** [comment]
What were the problems in the first place? (Would be nice to log in the next patch notes on what the problems were and that they were fixed)

**Comment by heinrich5991** [comment]
Wasting less network traffic, basically. The server info packets are smaller and the server info requests only need to be sent once. I should resurrect this PR somewhen.

**Comment by heinrich5991** [comment]
@east Is it okay to merge this without an extra roundtrip? We already throttle server info requests.

**Comment by east** [comment]
Does it support falling back to sending only meta information on flood? (Like the current implementation does)

**Comment by heinrich5991** [comment]
Yes.

**Comment by east** [comment]
Well, then I agree on the merge since that approach seems to do well.

**Comment by east** [comment]
It looks like there is a trailing zero missing in the old 'dtsf' response, my current client doesn't accept it.

**Comment by east** [comment]
That issue only occurs when the server is empty.

**Comment by heinrich5991** [comment]
Diff to previous version:
```diff
diff --git a/src/engine/server/server.cpp b/src/engine/server/server.cpp
index 95472fc73..4081bd0c3 100644
--- a/src/engine/server/server.cpp
+++ b/src/engine/server/server.cpp
@@ -1449,6 +1449,9 @@ void CServer::SendServerInfo(const NETADDR *pAddr, int Token, int Type, bool Sen
 
        RESET();
 
+       if(Type == SERVERINFO_64_LEGACY)
+               pp.AddInt(PlayersSent); // offset
+
        if(!SendClients)
        {
                SEND(pp.Size());
@@ -1461,9 +1464,6 @@ void CServer::SendServerInfo(const NETADDR *pAddr, int Token, int Type, bool Sen
                PrefixSize = sizeof(SERVERBROWSE_INFO_EXTENDED_MORE);
        }
 
-       if(Type == SERVERINFO_64_LEGACY)
-               p.AddInt(PlayersSent); // offset
-
        int Remaining;
        switch(Type)
        {
```

File: src/engine/server/server.cpp [hunk 2]
```diff
@@ -1380,76 +1384,154 @@ void CServer::SendServerInfo(const NETADDR *pAddr, int Token, bool Extended, int
...
 
-	for(i = 0; i < MAX_CLIENTS; i++)
+	int Remaining;
+	switch(Type)
+	{
+	case SERVERINFO_EXTENDED: Remaining = -1; break;
+	case SERVERINFO_64_LEGACY: Remaining = 24; break;
+	case SERVERINFO_VANILLA: Remaining = VANILLA_MAX_CLIENTS; break;
+	case SERVERINFO_INGAME: Remaining = VANILLA_MAX_CLIENTS; break;
+	}
```
Comment (east) [inline src/engine/server/server.cpp]: `src/engine/server/server.cpp:1498:16: warning: ‘Remaining’ may be used uninitialized in this function [-Wmaybe-uninitialized]`
I guess we should fix this warning

Comment (heinrich5991) [inline src/engine/server/server.cpp]: Fixed.

File: src/engine/server/server.cpp [hunk 3]
```diff
@@ -1380,76 +1384,155 @@ void CServer::SendServerInfo(const NETADDR *pAddr, int Token, bool Extended, int
...
+				if(Type == SERVERINFO_VANILLA || Type == SERVERINFO_INGAME)
+					break;
+
+				// Otherwise we're SERVERINFO_64_LEGACY.
+				SEND(pp.Size());
+				RESET();
+				p.AddInt(PlayersSent); // offset
+				Remaining = 24;
+			}
+			else if(Remaining > 0)
```
Comment (heinrich5991) [inline src/engine/server/server.cpp]: Oh damn. I believe this should be an `if` not an `else if`: The 49th player would get added to the second packet, but should go into the third one. Will fix this today.

Comment (heinrich5991) [inline src/engine/server/server.cpp]: Fixed.

**Comment by heinrich5991** [comment]
Diff:
```diff
diff --git a/src/engine/server/server.cpp b/src/engine/server/server.cpp
index 93c378373..bbee1254f 100644
--- a/src/engine/server/server.cpp
+++ b/src/engine/server/server.cpp
@@ -1494,7 +1494,7 @@ void CServer::SendServerInfo(const NETADDR *pAddr, int Token, int Type, bool Sen
                                p.AddInt(PlayersSent); // offset
                                Remaining = 24;
                        }
-                       else if(Remaining > 0)
+                       if(Remaining > 0)
                        {
                                Remaining--;
                        }
```

**Comment by heinrich5991** [comment]
```diff
diff --git a/src/engine/client/serverbrowser.cpp b/src/engine/client/serverbrowser.cpp
index 49a565afe..b5624241e 100644
--- a/src/engine/client/serverbrowser.cpp
+++ b/src/engine/client/serverbrowser.cpp
@@ -407,7 +407,7 @@ CServerBrowser::CServerEntry *CServerBrowser::Add(const NETADDR &Addr)
 
        // set the info
        pEntry->m_Addr = Addr;
-       pEntry->m_ExtraToken = rand() & 0xffff;
+       pEntry->m_ExtraToken = secure_rand() & 0xffff;
        pEntry->m_Info.m_NetAddr = Addr;
 
        pEntry->m_Info.m_Latency = 999;
diff --git a/src/engine/server/server.cpp b/src/engine/server/server.cpp
index 3ca026459..c104535a7 100644
--- a/src/engine/server/server.cpp
+++ b/src/engine/server/server.cpp
@@ -1491,7 +1491,7 @@ void CServer::SendServerInfo(const NETADDR *pAddr, int Token, int Type, bool Sen
                                // Otherwise we're SERVERINFO_64_LEGACY.
                                SEND(pp.Size());
                                RESET();
-                               p.AddInt(PlayersSent); // offset
+                               pp.AddInt(PlayersSent); // offset
                                Remaining = 24;
                        }
                        if(Remaining > 0)
```
